GitCommands/RevisionReader.cs

```diff
@@ -451,17 +439,55 @@ DateTime ParseUnixDateTime()
             return true;
         }
 
+        [CanBeNull]
+        private static (string body, string additionalData) ParseCommitBody([NotNull] IStringLineReader reader, [NotNull] string subject)
+        {
+            int lengthOfSubjectRepeatedInBody = subject.Length + 2/*newlines*/;
+            if (reader.Remaining == lengthOfSubjectRepeatedInBody + EndOfBody.Length)
+            {
+                return (body: subject, additionalData: null);
+            }
+
+            string tail = reader.ReadToEnd() ?? "";
+            int indexOfEndOfBody = tail.LastIndexOf(EndOfBody, StringComparison.InvariantCulture);
+            if (indexOfEndOfBody < 0)
+            {
+                // TODO log this parse error
+                Debug.Fail("Missing end-of-body marker in the log -- this should not happen");
+                return (body: null, additionalData: null);
+            }
+
+            string additionalData = null;
+            if (tail.Length > indexOfEndOfBody + EndOfBody.Length)
+            {
+                additionalData = tail.Substring(indexOfEndOfBody + EndOfBody.Length).TrimStart();
+            }
+
+            string body = indexOfEndOfBody == lengthOfSubjectRepeatedInBody
+                          ? subject : tail.Substring(0, indexOfEndOfBody).TrimEnd();
+            return (body, additionalData);
+        }
+
         public void Dispose()
         {
             _cancellationTokenSequence.Dispose();
         }
 
         #region Nested type: StringLineReader
 
+        public interface IStringLineReader
+        {
+            int Remaining { get; }
+
+            string ReadLine([CanBeNull] StringPool pool = null, bool advance = true);
+
+            string ReadToEnd(bool advance = true);
+        }
+
         /// <summary>
         /// Simple type to walk along a string, line by line, without redundant allocations.
         /// </summary>
-        private struct StringLineReader
+        private struct StringLineReader : IStringLineReader
```

Comment:
Please do not make this `struct` an interface. Doing so causes it to be boxed, which allocates on the heap. If you need external access, just make the type public.

Reply:
done